Header Authorization

EnigmaCurry opened the pull request:
This PR will track implementing #118 which will add header based authorization to apps that don't normally support being hosted behind an oauth proxy, or to provide an additional layer of authorization. 

DONE: 
 * Installed traefik plugin to act as middleware after traefik-forward-auth to filter approved usernames - https://github.com/EnigmaCurry/traefik-header-authorization
 * added whoami example

TODO:
* configure all the .env-dist, Makefile, and docker-compose.instance.yaml for all apps that benefit from t-f-a.

Fixes #125 
Fixes #128

EnigmaCurry commented:
I added a hardcoded example to whoami and tested that it works. This needs to be made configurable from the env file and the reconfigure_oauth script needs to be modified to ask for the list of the allowed users (gitea email addresses). Do you want to test it and take over since you already made that script @mcmikemn ?

mcmikemn commented:
> Do you want to test it and take over since you already made that script @mcmikemn ?

Sure! Working on it...

mcmikemn reviewed (changes requested):
In the .env-dist files, instead of saying "With Oauth2 authentication, you can configure which email addresses...", make it more clear that the basic authorization can be configured but is not part of Oauth2 authentication. Maybe "In addition to Oauth2 authentication, you can configure basic authorization by entering which email addresses..."

mcmikemn commented:
@EnigmaCurry, this looks good to me and I tested it with whoami and homepage.

EnigmaCurry commented:
I think it would be nicer if it it offered a trinary response "Do you want to enable authentication?" 1) No 2) HTTP Basic 3) Oauth2 

I know we have no examples of that in the makefiles up to this point. I can't imagine why anyone would mix basic and oauth at the same time.

EnigmaCurry commented:
well it could work like this:

Do you want to enable traefik authentication (in addition to any app auth)? N/y

If yes, detect if t-f-a is installed, and ask "Do you want to use Oauth2" and if no or not installed, ask "Do you want to use HTTP Basic Auth". 

No trinary response needed if its broken up into multiple y/n questions.

EnigmaCurry commented:
I moved the middleware config to the Traefik config, and added a script `make groups` to manage the group config.

I modified the whoami example with a hardcoded header-authorization-group middleware for the `users` group.

We still need a script that the app Makefile can run to select which kind of auth the user wants, and if they do want oauth2 it needs to set the correct middleware for the group. (To keep things simple, I think the question should be worded about what group the user wants to apply (ie. `users`)  not which middleware (ie. not `header-authorization-group-users`)

mcmikemn commented:
@EnigmaCurry, I'm adding the "remove user from group" function to `_scripts/reconfigure-header-authorization-groups`, but I can't get 1 line to work: https://github.com/EnigmaCurry/d.rymcg.tech/blob/67cffb179a54a78a71c4832994c458db85392bce/_scripts/reconfigure_header_authorization_groups#L125
It errors with "Cannot iterate over null."

I also added display info to the other functions to make this script easier to use. And I added "enter blank to skip" to the function calls, because otherwise the only way out of a command was to <ctrl>-C the script.

EnigmaCurry commented:
I couldn't figure out how to remove all the selected users in one line so I had to use a loop

https://github.com/EnigmaCurry/d.rymcg.tech/blob/1960e9504d6d835d666f6a93735a0891bb6fce89/_scripts/reconfigure_header_authorization_groups#L132-L135

mcmikemn commented:
I combined all auth config questions, made authroization group configurable in docker-compose.instance.yaml, added reconfigure_menu for general use (I didn't incporate it into `reconfigure_header_authorization_groups`, but it would be a good use for it), updated readmes.

I only modified and tested whoami - if this works well and is the way we want to go, I still need to update each app that will have oauth2 as an option (audiobookshelf, drawio, grocy, homepage, invidious, lemmy (by removing the option to add external authentication), photoprism, piwigo, wordpress).

EnigmaCurry commented on _scripts/reconfigure_auth:
```diff
@@ -0,0 +1,52 @@
+ENV_FILE=$1
+VAR_PREFIX=$2
+
+## note: should we add comments to reconfigure_htpasswd and reconfigure_oauth2 indicating they were designed to be called by this script, not independently?
```
We can set some crazy env var in this script like __D_RY_CONFIG_ENTRY=reconfigure_auth and then if that variable is not set correctly in reconfigure_htpasswd or reconfigure_oauth2, then it aborts.

EnigmaCurry replied:
yea I think thats the idea.

mcmikemn replied:
added.

EnigmaCurry commented:
Thanks mike! 

 * Please use four spaces instead of tabs in scripts.
 * Do not `source` scripts to return values. You need to print and parse instead. 
 * I would recommend using stderr for human text and stdout for the response text (which for multiple values, needs to be a JSON response, and parsed from the caller)

EnigmaCurry commented:
```
OPTIONS=$1
PROMPT=$2
DEFAULT=$3
MIN=${4:-1}
MAX={5:-$MIN}
check_var OPTIONS PROMPT DEFAULT
check_num MIN MAX
```
you can put something like this at the top to check the correct args are set before you start processing them.

EnigmaCurry commented:
I have merged the script-wizard branch into the header-authorization branch.

mcmikemn commented:
I updated `reconfigure_auth` to incorporate script-wizard. It's only functional in `whoami` at the moment. Once it's good, I'll update the other apps that benefit from t-f-a.

EnigmaCurry commented on whoami/.env-dist:
```diff
@@ -27,7 +27,7 @@ WHOAMI_HTTP_AUTH=
 # by entering which authorization group can log into your app. You create
-# groups of email addresses in the `traefik` folder by running `makr groups`. 
+# groups of email addresses in the `traefik` folder by running `make groups`. 
 WHOAMI_AUTHORIZED_GROUP=
```
I think this should be WHOAMI_OAUTH2_AUTHORIZED_GROUP

mcmikemn replied:
Fixed.

EnigmaCurry commented:
Added confirmation before installing script-wizard.

EnigmaCurry commented on _scripts/reconfigure_oauth2:
```diff
@@ -12,9 +12,13 @@ set -e
 shift
 
+## env var for authorized emails
+VAR2=${2}   
```
Something's wrong with the args here. You use `shift` (which pops the first arg off and the second arg becomes $1 for each time you call `shift`), so it shouldn't be $2 here. When I changed it to $1 and I inspected the value it was `VAR=default=yes` which also seems wrong.

EnigmaCurry replied:
These should not be called `VAR` and `VAR2` please pick descriptive names.

mcmikemn replied:
$2 after the shift was correct because $1 is used elsewhere, but I'll rename the vars

mcmikemn replied:
fixed in upcoming commit

EnigmaCurry commented on _scripts/reconfigure_oauth2:
```diff
@@ -35,9 +39,28 @@ if [[ ${1} =~ ^default= ]]; then
+    echo 
+    echo "Using OpenID/OAuth2 will require a login to access your app. You can configure basic authorization by assigning an authorization group (authorization groupscontain email addresses that are allowed to log into your app - you configure authorization groups by running \`make groups\` in the \`traefik\` directory)."
+    echo 
+    echo "SECURITY NOTE: Using OpenID/OAuth2 is on top of any authentication/authorization service your app provides. OpenID/Oauth2 will require a login to access your app and permit only specific logins, but it will not affect what a successfully logged-in person can do in your app. If your app has a built-in authorization mechanism that can check for the user header that traefik-forward-auth se ...
```
I think the security note is out of date assuming that we enforcing the groups, its not a concern anymore.

mcmikemn replied:
You want me to remove the entire paragraph? from `reconfigure_oauth2` (i.e., the prompt when you're starting to configure oauth2), from the app readmes, and from the t-f-a readme?

EnigmaCurry replied:
how about replacing it with:

OAuth2 uses traefik-forward-auth to delegate authentication to an external authority (eg. a self-deployed gitea instance). Accessing this app will require all users to login through that external service first. Once authenticated, you may be authorized access only if your login id matches the member list of the predefined security group configured for the app (`WHOAMI_OAUTH2_AUTHORIZED_GROUP`) . Security groups are defined in the Traefik config (`TRAEFIK_HEADER_AUTHORIZATION_GROUPS`) and can be modified using `make groups` (in the traefik directory).

EnigmaCurry replied:
> You want me to remove the entire paragraph? from `reconfigure_oauth2` (i.e., the prompt when you're starting to configure oauth2), from the app readmes, and from the t-f-a readme?

I think its probably good to have it in each app readme, but specifically we dont need to worry about telling the user that "any user can do whatever they want" because now they have to be a part of a group.

mcmikemn replied:
updated.

EnigmaCurry commented:
reconfigure_oauth2 now checks if traefik-forward-auth is installed.

EnigmaCurry commented on _scripts/reconfigure_oauth2:
```diff
@@ -9,22 +9,26 @@ BIN=$(dirname ${BASH_SOURCE})
 source ${BIN}/funcs.sh
 set -e
 
+if [[ __D_RY_CONFIG_ENTRY != reconfigure_auth ]]; then
```
${__D_RY_CONFIG_ENTRY}

mcmikemn replied:
fixed.

EnigmaCurry commented on _scripts/reconfigure_htpasswd:
```diff
@@ -9,6 +9,12 @@ BIN=$(dirname ${BASH_SOURCE})
 source ${BIN}/funcs.sh
 set -e
 
+if [[ __D_RY_CONFIG_ENTRY != reconfigure_auth ]]; then
```
${__D_RY_CONFIG_ENTRY}

mcmikemn replied:
fixed

mcmikemn commented on piwigo/README.md:
```diff
@@ -29,8 +19,30 @@ Alternatively to running `make config`, you can manually copy
  * `MARIADB_PASSWORD` the mysql user password
  * `TIMEZONE` the timezone, in the format of [TZ database name](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones)
  * `PIWIGO_OAUTH2` set to `yes` to configure Oauth2 authentication (in addition to Piwigo's own internal authentication)
+ * `PIWIGO_AUTHORIZED_IDENTITIES` enter a comma-separated list of email addresses authorized to log into this app (must match Gitea accounts)
```
this needs to reflect groups, not emails

mcmikemn replied:
fixed.

mcmikemn commented on traefik-forward-auth/README.md:
```diff
@@ -85,31 +100,40 @@ See the [whoami](../whoami/) app for examples.
 ## Logging out
 
-User logout is a multi-phase endevour:
+User logout is a multi-phase endeavour:
```
spelling: "endeavor"

mcmikemn reviewed (changes requested):
all app readmes refer to WHOAMI* env var - change that to refer to each app's env var.

EnigmaCurry commented:
> Do you want to rename `confirm.sh` to `confirm_script_wizard.sh`, and `confirm` to `confirm.sh`?

My intention was to replace the old bash `confirm` and `ask` with the new script-wizard equivalents. Which, the only nicer thing about them I can see is that it normalizes the printed response as `Yes` or `No` in blue text. (I already backported the green ? prompt.)

So i easily replaced confirm, but I still need the old one as confirmation before installing script-wizard in the first place.

I tried replacing ask but i couldn't replicate the "response on a new line" feature we have now, so i did not replace ask for now.

As far as the naming, I like confirm.sh because it makes it clear it is bash. It doesn't really matter which one you use they function nearly identically, but I would prefer to use the new one `confirm`. I hope its not confusing.

mcmikemn commented:
> > Do you want to rename `confirm.sh` to `confirm_script_wizard.sh`, and `confirm` to `confirm.sh`?
> 
> My intention was to replace the old bash `confirm` and `ask` with the new script-wizard equivalents. Which, the only nicer thing about them I can see is that it normalizes the printed response as `Yes` or `No` in blue text. (I already backported the green ? prompt.)
> 
> So i easily replaced confirm, but I still need the old one as confirmation before installing script-wizard in the first place.
> 
> I tried replacing ask but i couldn't replicate the "response on a new line" feature we have now, so i did not replace ask for now.
> 
> As far as the naming, I like confirm.sh because it makes it clear it is bash. It doesn't really matter which one you use they function nearly identically, but I would prefer to use the new one `confirm`. I hope its not confusing.

I understood all of that. My assumption was that we _had to_ keep the pre-script-wizard version of `confirm.sh` because we'd need to confirm the _installation_ of script-wizard. We could use `ask` for that but since we want to keep both versions of `confirm`, I was trying to name them so it wouldn't look like there was an old and a new version of the same script and thus one might get accidentally pruned - the name would indicate that one was to be used to confirm s-w installation and the other would be used _with_ s-w after it was installed.

EnigmaCurry commented:
I'm going to rewrite the `make groups` to use script-wizard

mcmikemn commented:
I think the comments you added to `confirm` and `confirm.sh` are good enough.

mcmikemn commented on _scripts/reconfigure_auth:
```diff
@@ -0,0 +1,57 @@
+unset_http_auth() {
+    ${BIN}/reconfigure ${ENV_FILE} "${VAR_PREFIX}_HTTP_AUTH"=
+    TMP_PASSWORD=$(mktemp)
+    if [[ -f passwords.json ]]; then
```
This check should also me made in `reconfigure_htpasswd`

mcmikemn commented on traefik/README.md:
there are some lines in here about configuring the traefik dashboard and using HTTP Basic Auth - these should be updated to also mention Oauth2.

mcmikemn replied:
nope

mcmikemn commented on traefik/config/setup.sh:
```diff
@@ -43,7 +44,7 @@ ytt_template() {
         -v dashboard_entrypoint_host="${TRAEFIK_DASHBOARD_ENTRYPOINT_HOST}" \
         -v dashboard_entrypoint_port="${TRAEFIK_DASHBOARD_ENTRYPOINT_PORT}" \
-        -v dashboard_auth="${TRAEFIK_DASHBOARD_AUTH}" \
+        -v dashboard_auth="${TRAEFIK_DASHBOARD_HTTP_AUTH}" \
```
This needs to also be able to handle Oauth2.

mcmikemn replied:
nope

mcmikemn commented on traefik/docker-compose.yaml:
```diff
@@ -72,7 +73,7 @@ services:
       - TRAEFIK_DASHBOARD_ENTRYPOINT_HOST=${TRAEFIK_DASHBOARD_ENTRYPOINT_HOST:-127.0.0.1}
       - TRAEFIK_DASHBOARD_ENTRYPOINT_PORT=${TRAEFIK_DASHBOARD_ENTRYPOINT_PORT:-8080}
-      - TRAEFIK_DASHBOARD_AUTH
+      - TRAEFIK_DASHBOARD_HTTP_AUTH
```
I'm pretty sure that `TRAEFIK_DASHBOARD_OAUTH2` needs to be included here, too

mcmikemn replied:
nope

mcmikemn reviewed (commented):
At this point, I think all that's missing is the re-write of `reconfigure_header_authorization_groups` to work with script-wizard.

mcmikemn reviewed (commented):
At this point, I think all that's missing is the re-write of `reconfigure_header_authorization_groups` to work with script-wizard.

EnigmaCurry commented:
`make groups` is now rewritten for script-wizard, I'm sure theres bugs, bash is still not pretty even with the helper, please test for bugs. 

make sure to grab the latest script-wizard by re-running `make install-script-wizard` in the root dir.

mcmikemn commented:
Great improvement, @EnigmaCurry!

mcmikemn commented on _scripts/reconfigure_header_authorization_groups:
```diff
@@ -0,0 +1,250 @@
+    echo "Enter the new user id(s) to add, one per line:"
+    while :
+    do
+        user_id=$(wizard ask "Enter a user ID (press Esc or enter a blank value to finish)" --allow-blank)
```
`--allow-blanks` isn't working. 
```
> Security Group Manager (main menu): Group Manager
> Choose a group to manage imagegenerators
> Managing group imagegenerators :: select a task :: Add members
Enter the new user id(s) to add, one per line:
error: unexpected argument '--allow-blank' found

tip: to pass '--allow-blank' as a value, use '-- --allow-blank'

Usage: script-wizard ask <QUESTION|DEFAULT|--json>

For more information, try '--help'.
Wrote new TRAEFIK_HEADER_AUTHORIZATION_GROUPS to ../traefik/.env_server01.thewooskeys.com_default
```

mcmikemn replied:
fixed with version 0.1.19 of script-wizard

mcmikemn commented on _scripts/reconfigure_header_authorization_groups:
```diff
@@ -0,0 +1,250 @@
+        echo "Enter the new user id(s) to add, one per line:"
+        while :
+        do
+            user_id=$(wizard ask "Enter a user ID (Press Esc or enter a blank value to finish)" --allow-blank)
```
`--allow-blanks` isn't working. 
```
> Security Group Manager (main menu): Group Manager
> Choose a group to manage imagegenerators
> Managing group imagegenerators :: select a task :: Add members
Enter the new user id(s) to add, one per line:
error: unexpected argument '--allow-blank' found

tip: to pass '--allow-blank' as a value, use '-- --allow-blank'

Usage: script-wizard ask <QUESTION|DEFAULT|--json>

For more information, try '--help'.
Wrote new TRAEFIK_HEADER_AUTHORIZATION_GROUPS to ../traefik/.env_server01.thewooskeys.com_default
```

mcmikemn replied:
fixed with version 0.1.19 of script-wizard

EnigmaCurry commented:
I've tested all the apps you configured, they work great!

There's still apps though that aren't configured, and are broken because they can't call `reconfigure_htpasswd`.

EnigmaCurry commented:
I have used `true` in most of the other .env files so I am changing all the OAUTH2=yes to OAUTH2=true.

EnigmaCurry commented:
we should have a utility that outputs all the urls you are expected to paste into your gitea app config. it could scan for all the running apps that have the middleware attached and print the list of all of them, then you just copy and paste to gitea.

EnigmaCurry commented:
I started going alphabetically through the ones I think we still need to fix, I got to nextcloud before I stopped.

mcmikemn commented:
I tried to instance-iate nextcloud but couldn't make it work. Then I switched to the master branch and simply tried to install it, and I can't make _that_ work.

EnigmaCurry commented:
I'm gonna skip nextcloud and only focus on the ones that currently have basic auth

EnigmaCurry commented:
@mcmikemn  I have gone through all the apps that had http basic auth, and upgraded them to use reconfigure_auth. 

I fixed traefik-forward-auth to use auth host mode, so when you create the Gitea oauth application, you only need to enter a single redirect uri : `https://auth.example.com/_oauth`. 

`make groups` is renamed to `make sentry`. This includes a new method to print all the redirect uris, but this shouldn't be necessary as long as auth host mode is working.

Please test creating a new gitea oauth app, reconfigure/redeploy traefik-forward-auth, and then retest a handful of these applications, and carefully read all the options, and see if it a) works and b) makes sense.

EnigmaCurry commented:
I upgraded script-wizard again, so run `make install-script-wizard` again in the root project dir to upgrade to v0.1.20.

mcmikemn commented on vaultwarden/README.md:
````diff
@@ -41,7 +45,7 @@ registration:
 ```
 # Allow public registrations, but don't forget to disable this again later.
-make enable-registration
+gmake enable-registration
````
Is this supposed to be `gmake` or `make`?

EnigmaCurry replied:
make :+1:

EnigmaCurry commented:
I removed the `make install-script-wizard` target, and now it checks the ${ROOT_DIR}/.tools.lock.json file for the required locked version instead, and if the version is not installed, it runs the installer after confirmation.
